### ISSUE TITLE
Allow Schema Value to be Owned

### ISSUE THREAD

**Issue by kayyagari** [Stranger6667/jsonschema-rs#145]
In the current form it is impossible to retain a non-static reference to `JSONSchema` until the end of the program's execution. The below block demonstrates the problem.

```rust
#[actix_web::main]
async fn main() -> std::io::Result<()> {
    let schema_file = fs::File::open("config/schema.json").unwrap();

    let schema: Value = serde_json::from_reader(schema_file).unwrap();
    let validator = JSONSchema::compile(&schema, None).unwrap();

    let ad = AppData{
        validator: Arc::new(validator)
    };

    HttpServer::new(move ||{
        App::new()
            .data(ad.clone())
            .service(echo)
    })
    .bind("localhost:9070")?
    .run()
    .await
}

#[derive(Clone)]
pub struct AppData<'a> {
    pub validator: Arc<JSONSchema<'a>>
}
```
The above code results in the below compiler error:

```text
error[E0597]: `schema` does not live long enough
  --> src/main.rs:15:41
   |
15 |     let validator = JSONSchema::compile(&schema, None).unwrap();
   |                     --------------------^^^^^^^-------
   |                     |                   |
   |                     |                   borrowed value does not live long enough
   |                     argument requires that `schema` is borrowed for `'static`
...
31 | }
   | - `schema` dropped here while still borrowed
```

One(or the only?) way to get this fixed is to let `JSONSchema` own the `Value` containing schema.

**Comment by kyle-mccarthy** [comment]
@kayyagari I know that this doesn't address making the schema value owned, but your example is probably a good use case for [Box::leak](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.leak). In this case leaking the memory doesn't matter since you want the value to exist for the duration of the program.

```rust
let schema: Value = serde_json::from_reader(schema_file).unwrap();
let schema_boxed: &'static Value = Box::leak(Box::new(schema));
let validator = JSONSchema::compile(&schema_boxed).unwrap();
```

**Comment by lichr** [comment]
I have encountered same issue, tried to workaround the issue by lying to the borrow checker about the lifetime of input json Value like this: 
```rust
/// IMPORTANT: this struct must not be moved
/// this is a self referential struct: JSONSchema holds a reference to doc
pub struct Schema<'a> {
    // step1: we need to move doc here, so that it's address is fixed
    pub doc: Value,
    // step2: we create schema with self.doc, and move it here
    pub compiled: Option<JSONSchema<'a>>,
}

impl<'a> Schema<'a> {
    fn compile(&mut self, options: &CompilationOptions) {
        // IMPORTANT: here we lie about doc's lifetime
        let doc: &'static Value = unsafe { std::mem::transmute(&self.doc) };
        let compiled = options.compile(doc).unwrap();
        self.compiled = Some(compiled);
    }
}

```

What I was doing is to load many schemas and put them into a hashmap:
```rust

{
    schemas: HashMap<String, Box<Schema<'a>>>,
}

// at some point:
            let mut schema = Box::new(Schema {
                compiled: None,
                doc: doc.clone(),
            });
            schema.compile(&self.options);
            self.schemas.insert(name.to_string(), schema);

```

I think one possible way to improve this is to let the compile method 'consume' the json Value so that the resulted JSONSchema doesn't hold a reference to the input json Value, this could make things much simpler.

**Comment by johnsom** [comment]
I also hit this issue and banged my head against the wall on it. I'm learning Rust and thought I was doing something stupid.
I agree, having compile consume the Value would make using this more obvious.
Thanks for the crate!

**Comment by elbaro** [comment]
A common use case is to compile a schema once and reuse it for validation multiple times.
In this case, we don't need to compile fast by avoiding the copy.
In fact, I guess the owned version will validate faster due to memory locality.

**Comment by luke-biel** [comment]
I would love this to be a feature. It's a mess when trying to create a service that validates multiple object types with caching built in.

**Comment by robot-head** [comment]
Also running into this exact issue building exactly this:

> I would love this to be a feature. It's a mess when trying to create a service that validates multiple object types with caching built in.

Is the official answer to do a `Box::leak`? I want to put this inside a OnceCell that pre-caches the schemas at service startup

**Comment by alexjg** [comment]
This seems like it would also be useful for `ValidationError`s. Currently implementing any kind of type which wraps a validation error requires holding on to the input reference somewhere, this can be irritating if you're doing something like accumulating errors over many instances which are each transient.

**Comment by Stranger6667** [comment]
Folks, I completely agree with the mentioned points and would like to have it changed this way.

Maybe I overestimated the impact (compilation usually happens much less often than validation), and it will be alright to clone the input schema into `resolver` [under the hood](https://github.com/Stranger6667/jsonschema-rs/blob/master/jsonschema/src/compilation/options.rs#L118). If anybody knows any good way to have it more compact, I'd be happy to learn about it.

At the moment, I don't have much bandwidth to work on this, but I will happily review a PR that brings this change.

**Comment by alexjg** [comment]
I can take this on. My main question is do we care about API breakage?

**Comment by Stranger6667** [comment]
@alexjg 

Awesome! I think that with this kind of change, the breakage is inevitable and is nicely motivated by the comments above. However, it would be nice to reduce the breakage surface if we can, but it should not block us from this kind of usability improvement.

**Comment by Stranger6667** [comment]
Solved in #255 by @alexjg ! :) Thank you very much for implementing this :)

**Comment by Stranger6667** [comment]
Released in `0.12.0`
